cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -1711,7 +1711,7 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
                                 ));
                                 ctx.emit(Inst::xmm_mov(
```

Comment:
Looking at the pattern matching and such, it looks like the codegen here is actually the same as-if the specialized pattern matching in this function weren't present? Basically it looks like this is generating `palignr; pmovsxbw` for each lhs/rhs, and then it's `pmullw` for the result. That looks like the same codegen as if the component operations of `i16x8.extmul_high_i8x16_s` were each translated individually.

If that's the case is it necessary to have a specialized pattern-match here for x64? E.g. were the wasm instructions primarily for other platforms which have a native instruction for the operation?

Reply:
Also, how come a temp register is used? It seems like the goal of `palignr` is to get the high 8 values into the low 8 values, and then the `pmovsxbw` will sign-extend the low values into 16-bit values, but looking at the intructions it seems like that can all be done with the same register?

Reply:
> Looking at the pattern matching and such, it looks like the codegen here is actually the same as-if the specialized pattern matching in this function weren't present? Basically it looks like this is generating `palignr; pmovsxbw` for each lhs/rhs, and then it's `pmullw` for the result. That looks like the same codegen as if the component operations of `i16x8.extmul_high_i8x16_s` were each translated individually.
> 
> If that's the case is it necessary to have a specialized pattern-match here for x64? E.g. were the wasm instructions primarily for other platforms which have a native instruction for the operation?

@alexcrichton Another good observation. Yes, obviously this instruction is just the combination of a 3 instruction sequence written as one. For arm these 3 instructions can be lowered efficiently more easily with this 3 instruction combination for all 3 variants, i.e.  (i16x8.extmul_high_i8x16_s, i32x4.extmul_high_i16x8_s, i64x2.extmul_high_i32x4_s) so the lowering is optimal but for x64 more optimal lowering is only available for a few of the variants. Namely i16x8.extmul_high_i8x16_s and i16x8.extmul_high_i8x16_u does not have a more optimal lowering so I end up just inlining the codegen here. In fact the mistake was simply a copy paste error in that inlining, not seeing a way to print the final codegen to triple-check,  and then not having adequate to testing to catch the mistake. For the other variants the specialized codegen I think saves an instruction or two (though I haven't counted explicitly).

Reply:
Ah ok! Do you think it'd be possible to structure the code such that it falls-through to the normal lowerings for non-specialized implementations, but specialized implementaitons (the other widths > 8) still have pattern-matched codegen? Either way seems fine to me though

Reply:
Hi @alexcrichton, I think so but I am not quite confident how to write this ... particular when there are two input registers here. What you gain is the normal benefit of code reuse but the lowering should be no more efficient than what you see here and may introduce a few more jumps. If you are ok with either way I'd prefer to just merge this fix now and refactor after a discussion with @cfallin.

Reply:
Oh sure yeah, I think it's fine to land this now and refactor later (although I think some tests here would still be good)

Reply:
FWIW I'm fine with this minor code duplication for now; a bigger refactor later when we have a more systematic way of writing lowering pattern-matching seems reasonable.

Reply:
Hi @alexcrichton, did you have in many any particular type of test in mind? There is spec tests, file tests, unit tests, and in the future fuzz testing but for proving correctness I think the spec tests have been what we've used. The unit tests help confirm encodings as long as we are confirming with as or some other tool.